## QTL mapping
This pipeline is based on [R/qtl](https://rqtl.org/) package ([R/qtl2](https://kbroman.org/qtl2/index.html) is available now).

---

* **install R/qtl**
```
  install.packages("qtl")
```
* **R/qtl input**

R/qtl needs a special input format. After calling variants by general ways, like GATK, we need to transfer VCF format to R/qtl input format.

1. Call variants

Here, an example of calling variants by GATK UnifiedGenotyper is given. 

```
  # "samples.bam.list" gives bam file paths, one row per sample. Parental samples are included in it.
  Samples=`cat samples.bam.list | xargs -I BAM_FILE echo -n "-I BAM_FILE "` 
  
  # Call variants by GATK UnifiedGenotyper
  java -jar GenomeAnalysisTK.jar \
      -R IRGSP-1.0_genome.fasta \
      -T UnifiedGenotyper -glm BOTH -nt 20 \
      -stand_call_conf 30.0 -stand_emit_conf 30.0 -rf MappingQuality -mmq 20 \
      -o samples.ug.vcf \
      ${Samples} \
      > samples.ug.log 2>&1
  
  # Compress VCF file and index. This step is optional, while zip file is required for several softwares and is space-saving.
  bgzip samples.ug.vcf > samples.ug.vcf.gz
  tabix -p vcf samples.ug.vcf.gz
```

2. Identify markers

Markers mean ploymorphyism sites (could be SNPs and Indels) between parental lines. Proper markers are important for downstream analysis.
```
  bgzip -dc samples.ug.vcf.gz | \
      perl vcf2markers.pl \
          -i - \
          --segregating ParentA,ParentB --out-symbols A,B --method m1 --remove-sample Sample1 \
          --max-alleles 2 --min-alleles 2 --minor-allele-frequence 0.2 --max-missing-proportion 0.5 \
          > samples.ug.markers.vcf
```
3. Formating

Transfer chromosome positions (e.g. Chr01 1000) to genetic distances (cM) and combine markers with phenotype data. Recombination frequency is needed, for example, there are around 31 CO events per genome in rice (Si et al., 2015, New phyt). We can also roughly calculate recombination rate by "[vcf_process.pl](https://github.com/wl13/BioScripts)".
```
  perl vcf2rqtl.pl -v samples.ug.markers.vcf \
      -g IRGSP-1.0_genome.fasta.fai \
      -p samples.phenotype.txt \
      -m samples.m1.rqtl.pos.csv \
      -rt 31 -sy A,B | sed 's/-//g' \
      > samples.m1.rqtl.csv

  # An example for phenotype data (tab-separated):
  # #Sample	HD
  # sample1 67
  # sample2 90
```
* **QTL-mapping analysis**

There are many useful models and strategies in R/qtl package. Here, three methods, HK, EM and IMP, are listed:
1. Calculate LOD value by R/qtl
```
  Rscript qtl_analysis.R \
      QTL_outdir/ \ # directory of ouput files.
      samples.m1.rqtl.csv \
      samples.hd.scanone.hk.txt \
      samples.hd.scanone.em.txt \
      samples.hd.scanone.imp.txt
```

2. Count LOD values by slide windows (smooth-LOD way)

```
  # Transfer genetic distance to chromosome position
  perl qtl_marker2pos.pl -rt 31 \
      -pos samples.m1.rqtl.pos.csv \
      -cm samples.hd.scanone.hk.txt \
      > samples.hd.scanone.hk.pos

  # Count LOD value by slide-windows and plot smooth-LOD images along chromosomes.
  for window in "w10k" "w50k" "w100k" "w200k" "w500k" "w1m";
  do
      echo ${window}
      awk 'BEGIN{OFS="\t"} !/^#/ {print $2,$3,$4}' samples.hd.scanone.hk.pos | \
          perl smooth_window_count.pl -s - \
              -b IRGSP-1.0_genome.${window}.bed --methods average | awk 'BEGIN{OFS="\t"} !/^##/ {print $0}' | sed 's/#//' \
              > samples.hd.scanone.hk.pos.${window}.stats
      Rscript plot_smooth_windows.LOD.R \
          samples.hd.scanone.hk.pos.${window}.stats \
          /home/jxq/Data/rice/ref/centromere.bed \
          7.03 7.96 Soomth_LOD samples.hd.scanone.hk.pos.${window}.stats \
          samples.hd.scanone.hk.pos.${window}.stats.pdf
  done

  # Note:
  # "IRGSP-1.0_genome.${window}.bed" files could be generated by "bedtools makewindows".
  # 7.03 and 7.96 are LOD thresholds (1000 permutations) of 0.05 and 0.01, respectively. (find in "Permutation.tests.txt" file)
```
3. Generate high-LOD intervals

High-LOD markers with LOD value > 0.05 or 0.01 LOD thresholds. Confidence intervals of high-LOD regions are usually 1.8-LOD down.
```
  # Extand blocks to credible markers
  awk 'BEGIN{OFS="\t"} !/^#/ {if($4 >= 7.03){print $0;}}' samples.hd.scanone.hk.pos | \ # 7.03 is 0.05 LOD thresholds
      perl high_lod_site2intervals.pl -i - \
          -s samples.hd.scanone.hk.pos -d 1.8 \
          > samples.hd.scanone.hk.pos.lod1_8.extand.blocks

  # Merge overlapped regions
  awk 'BEGIN{OFS="\t"} !/Chrom/ {print $2,$5-1,$6;}' samples.hd.scanone.hk.pos.lod1_8.extand.blocks | \
      sort -nk 1.4,1 -nk 2,2 | bedtools merge -i - \
      > samples.hd.scanone.hk.pos.lod1_8.extand.blocks.bed

  # find genes located in the high-LOD regions
  bedtools intersect -a samples.hd.scanone.hk.pos.lod1_8.extand.blocks.bed \
      -b MSU_osa1r7_all.mod.gff3.genes.csv -wa -wb \
      > samples.hd.scanone.hk.pos.lod1_8.extand.blocks.gene.csv
```

